src/main/java/com/google/sps/servlets/BlobHost.java

```diff
@@ -32,17 +41,73 @@ public void doGet(HttpServletRequest request, HttpServletResponse response)
 
     String blobKeyString = request.getParameter("blobkey");
     if (DataUtils.isEmptyParameter(blobKeyString)) {
-        throw new IOException("You do not have access to that resource.");
+      throw new IOException("You do not have access to that resource.");
     }
-    BlobKey blobKey = new BlobKey(blobKeyString);
 
-    // TODO(dtjanaka@):
-    // check for access before serving image from blobkey
-    // note: blobkeys are not guessable
+    Entity assetEntity = new Entity("Image");
+    boolean isMask = false;
+
+    // Find asset with the given blobkey
+    try {
+      assetEntity = getAssetEntity(DataUtils.IMAGE, blobKeyString);
+    } catch (Exception e) {
+      assetEntity = getAssetEntity(DataUtils.MASK, blobKeyString);
+      isMask = true;
+    }
+
+    String projId = new String();
+    if (isMask) { // if Mask, parent's parent --> Project
+      try {
+        projId = (String)datastore
+                     .get(datastore.get(assetEntity.getParent()).getParent())
+                     .getProperty("proj-id");
+      } catch (Exception e) {
+        throw new IOException(
+            "You do not have permission to access this project.");
+      }
+    } else { // if Image, get parent --> Project
+      try {
+        projId = (String)datastore.get(assetEntity.getParent())
+                     .getProperty("proj-id");
+      } catch (Exception e) {
+        throw new IOException(
+            "You do not have permission to access this project.");
+      }
+    }
```

Comment:
There's a bunch of duplicated code between the two branches. How about putting the `if` inside of the `try`? Or make a variable for the object whose proj-id you are going to access.